### ISSUE TITLE
istio.VERSION references non-existent proxy image with locally committed changes

### ISSUE THREAD

**Issue by ayj** [istio/istio#4022]
The e2e makefile test targets (e.g. `make e2e_pilot`) don't work reliability with locally committed changes. The `generate_yaml` make target updates `PROXY_TAG` in `istio.VERSION` with the current repo's latest sha. `PROXY_TAG` is then used by the docker image target to download a prebuilt proxy image from a GCS bucket. The problem is that `PROXY_TAG` corresponds to the current repo (istio/istio) but that image hasn't been built/pushed from istio/proxy repo.

Minimal steps to reproduce the problem:

```bash
$ make generate_yaml
./install/updateVersion.sh -a jasonayoung,09c368f87d02499943324746aa48b20c5649e77d >/dev/null 2>&1

$ make docker.proxy
mkdir -p /home/jasonyoung/work/out/linux_amd64/release/docker_temp
bin/gobuild.sh /home/jasonyoung/work/out/linux_amd64/release/pilot-agent istio.io/istio/pkg/version ./pilot/cmd/pilot-agent

real    0m5.239s
user    0m12.820s
sys     0m1.132s
cp /home/jasonyoung/work/out/linux_amd64/release/pilot-agent /home/jasonyoung/work/out/linux_amd64/release/docker_temp
make: *** No rule to make target '/home/jasonyoung/work/out/linux_amd64/release/envoy-09c368f87d02499943324746aa48b20c5649e77d', needed by '/home/jasonyoung/work/out/linux_amd64/release/docker_temp/envoy'.  Stop.
```

One workaround is to explicitly set `PROXY_TAG` to a known good value, e.g.

```bash
# Save `PROXY_TAG` before it's overwritten by other make targets and 
# download the corresponding proxy image from GCS via `make init`
git checkout istio.VERSION
PROXY_TAG=$(sed 's/export PROXY_TAG="\([0-9a-z]*\)"/\1/;t;d' istio.VERSION)
make init
```

```
make generate_yaml
PROXY_TAG=$PROXY_TAG make docker.proxy
```

**Comment by mattdelco** [comment]
There's conflicting interpretations across istio for whether PROXY_TAG refers to the envoy image generated in the proxy repo or the proxy image built under the pilot subdir of the istio repo.  It's been in conflicted use for awhile but misc changes have been compounding on top of the confusion.  I'm not entirely sure which i correct (granted the one from proxy repo is mostly discontinued), or whether these should be split out.  I'll pick out Nathan since I was last aware of the change he made that built upon the confusion, and Andy to see if there's somebody to dispatch the task up to (or push up the dev oversight hierarchy [if such a thing exists]).

**Comment by mattdelco** [comment]
The quickest workaround might be to split out PROXY_TAG as ENVOY_TAG. Basically change istio.VERSION using:
<pre>
- export PROXY_TAG="80ec4f6c2a672b00e55d34ef9b5d26a42a98da42"
+ export ENVOY_TAG="80ec4f6c2a672b00e55d34ef9b5d26a42a98da42"
+ export PROXY_TAG="replace-by-actual-tag"
</pre>
then change Makefile and istio.deps to use ENVOY_TAG instead of PROXY_TAG.

There's also another subtle issue shown by the quoted error from make.  Given these rules in istio-docker.mk:
<pre>
$(ISTIO_DOCKER)/envoy: ${ISTIO_ENVOY_RELEASE_PATH} | $(ISTIO_DOCKER); cp ${ISTIO_ENVOY_RELEASE_PATH} $(ISTIO_DOCKER)/envoy
$(ISTIO_DOCKER)/envoy-debug: ${ISTIO_ENVOY_DEBUG_PATH} | $(ISTIO_DOCKER); cp ${ISTIO_ENVOY_DEBUG_PATH} $(ISTIO_DOCKER)/envoy-debug
</pre>
make needs to know how to make ${ISTIO_ENVOY_RELEASE_PATH} and ${ISTIO_ENVOY_DEBUG_PATH}, e.g., by adjusting this in the Makefile:
<pre>
- ${ISTIO_OUT}/envoy: init
+ ${ISTIO_ENVOY_RELEASE_PATH} ${ISTIO_ENVOY_DEBUG_PATH} ${ISTIO_OUT}/envoy: init
</pre>

**Comment by nmittler** [comment]
FYI @ZackButcher ... this may be related to some issues you've been having

**Comment by rkpagadala** [comment]
#3964 fixes this issue

**Comment by ldemailly** [comment]
name wise btw ENVOY_TAG should maybe rather be PROXY_BIN_URL as it's not a (docker nor git) tag
anyway and stays a url + SHA, and the other one PROXY(_IMAGE)_TAG

(because it's not just envoy, it's envoy + istio's mixerclient)

**Comment by rkpagadala** [comment]
fixed by #4062
